### PR TITLE
Capture location

### PR DESCRIPTION
**Pull request by aman-alfresco**

### CODE REVIEW

**Comment by agologan** [comment]
Please run spotless `./gradlew spotlessApply` and maintaining proper code indentation.

File: auth/src/main/kotlin/com/alfresco/auth/activity/LoginViewModel.kt [hunk 1]
```diff
@@ -37,6 +37,7 @@ class LoginViewModel(private val applicationContext: Context, authType: AuthType
     val onSaveSettings: LiveEvent<Int> = _onSaveSettings
     val isLoading = MutableLiveData<Boolean>()
     val identityUrl = MutableLiveData("")
+//    val identityUrl = MutableLiveData("mobileapps.envalfresco.com")
```
Comment (agologan) [inline auth/src/main/kotlin/com/alfresco/auth/activity/LoginViewModel.kt:40]: Remove comment.

File: capture/build.gradle [hunk 2]
```diff
@@ -49,4 +49,8 @@ dependencies {
     implementation libs.androidx.constraintlayout
 
     implementation libs.mavericks
+
+    implementation 'com.google.android.gms:play-services-location:18.0.0'
+    implementation 'androidx.exifinterface:exifinterface:1.3.2'
```
Comment (agologan) [inline capture/build.gradle:54]: Add dependencies to gradle dependency catalog (**libs.versions.toml**)

File: capture/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -5,9 +5,14 @@
 
     <uses-permission android:name="android.permission.CAMERA" />
     <uses-permission android:name="android.permission.RECORD_AUDIO" />
+    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
 
     <uses-feature android:name="android.hardware.camera.any" />
 
+    <!-- Required only when requesting background location access on Android 10 (API level 29). -->
+    <uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" />
```
Comment (agologan) [inline capture/src/main/AndroidManifest.xml:14]: As far as I can tell we are not using background location services so please remove this permission.

File: capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt [hunk 4]
```diff
@@ -1,15 +1,15 @@
...
+import android.location.Location
 import android.net.Uri
 import android.os.Bundle
-import android.view.KeyEvent
-import android.view.LayoutInflater
-import android.view.MotionEvent
-import android.view.View
-import android.view.ViewGroup
+import android.os.Looper
+import android.view.*
```
Comment (agologan) [inline capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt:12]: Prefer using single import instead of wildcard.

File: data/objectbox-models/default.json.bak [hunk 5]
```diff
@@ -0,0 +1,90 @@
+{
```
Comment (agologan) [inline data/objectbox-models/default.json.bak:1]: *.bak file is a backup and you don't need to check into VCS

File: data/build.gradle [hunk 6]
```diff
@@ -29,6 +29,5 @@ dependencies {
     api libs.androidx.work
 
     implementation libs.okhttp.logginginterceptor
-
```
Comment (agologan) [inline data/build.gradle:32]: Unnecessary change for this PR.

File: listview/src/main/kotlin/com/alfresco/content/listview/ListFragment.kt [hunk 7]
```diff
@@ -150,6 +150,7 @@ abstract class ListFragment<VM : ListViewModel<S>, S : ListViewState> :
         } else if (state.entries.isNotEmpty()) {
             state.entries.forEach {
                 if (it.type == Entry.Type.GROUP) {
+
```
Comment (agologan) [inline listview/src/main/kotlin/com/alfresco/content/listview/ListFragment.kt:153]: Unnecessary change for this PR.

File: auth/src/main/kotlin/com/alfresco/auth/activity/LoginViewModel.kt (hunk 1 above)
Comment (aman-alfresco) [reply auth/src/main/kotlin/com/alfresco/auth/activity/LoginViewModel.kt:40]: Done

File: capture/build.gradle (hunk 2 above)
Comment (aman-alfresco) [reply capture/build.gradle:54]: Done

File: capture/src/main/AndroidManifest.xml (hunk 3 above)
Comment (aman-alfresco) [reply capture/src/main/AndroidManifest.xml:14]: Done

File: capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt (hunk 4 above)
Comment (aman-alfresco) [reply capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt:12]: Done

File: gradle/libs.versions.toml [hunk 8]
```diff
@@ -103,3 +103,5 @@ spotless = "com.diffplug.spotless:spotless-plugin-gradle:5.12.5"
 subsamplingimageview = "com.davemorrissey.labs:subsampling-scale-image-view:3.10.0"
 
 timber = "com.jakewharton.timber:timber:4.7.1"
+
+play-location = "com.google.android.gms:play-services-location:18.0.0"
```
Comment (agologan) [inline gradle/libs.versions.toml:107]: be more explicit name it `google-playservices-location`
also this file is sorted alphabetically and grouped by context

**Comment by agologan** [comment]
Some of the items we discussed in Slack so there's traceability on the changes:
- attaching location information is optional
- location information doesn't need to be precise, so in the context of [Android 12](https://developer.android.com/about/versions/12/approximate-location) going to approximate is ok
- business logic should always be driven by the viewModel not by the fragment/activity/view

File: capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt (hunk 4 above)
Comment (agologan) [reply capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt:11]: This is still wildcard. How is it changed?

File: capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt [hunk 9]
```diff
@@ -176,7 +182,7 @@ class CameraFragment : Fragment(), KeyHandler, MavericksView {
         layout.viewFinder.controller = cameraController
 
         // Observe zoom changes
-        cameraController?.zoomState?.observe(this) {
+        cameraController?.zoomState?.observe(viewLifecycleOwner) {
```
Comment (agologan) [inline capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt:185]: This change has nothing to do with this PR.

File: capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt [hunk 10]
```diff
@@ -258,7 +264,18 @@ class CameraFragment : Fragment(), KeyHandler, MavericksView {
...
+        val metadata = ImageCapture.Metadata()
+
+        val location = Location("")
+
+        location.longitude = viewModel.longitude.toDouble()
+        location.latitude = viewModel.latitude.toDouble()
+
+        metadata.location = location
+
+        val outputOptions = ImageCapture.OutputFileOptions.Builder(photoFile).setMetadata(metadata).build()
```
Comment (agologan) [inline capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt:277]: This logic belongs in the viewModel.

**Comment by agologan** [comment]
Some notes on requesting location permission:
- recommend you see the [codelab](https://developer.android.com/codelabs/approximate-location) on approximate location 
- requesting permissions should be done via `PermissionFragment.requestPermissions`
- when dealing with an optional permission failure prompts and/or request rationale **shouldn't** be shown during normal user flow
- the helper does not currently deal with **optional** aspect of a permission request so it should be enhanced to do so - more explicitly it shouldn't show any prompts as explained previously if the user denies the permission

**Comment by agologan** [comment]
What I previously mentioned about ACCESS_FINE_LOCATION with PRIORITY_HIGH_ACCURACY I meant more along the lines of dialling down the priority to maybe PRIORITY_BALANCED_POWER_ACCURACY and consider requesting ACCESS_COARSE_LOCATION in the context of Android 12.

File: capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt (hunk 10 above)
Comment (aman-alfresco) [reply capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt:277]: Done

File: capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt (hunk 9 above)
Comment (aman-alfresco) [reply capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt:185]: Ok. I will undo this

File: capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt (hunk 4 above)
Comment (aman-alfresco) [reply capture/src/main/kotlin/com/alfresco/capture/CameraFragment.kt:11]: Done.

File: data/build.gradle (hunk 6 above)
Comment (aman-alfresco) [reply data/build.gradle:32]: It's some space change. you may ignore this

File: listview/src/main/kotlin/com/alfresco/content/listview/ListFragment.kt (hunk 7 above)
Comment (aman-alfresco) [reply listview/src/main/kotlin/com/alfresco/content/listview/ListFragment.kt:153]: Done. Removed the space

**Review by agologan — CHANGES_REQUESTED** [review]
_see inline and in the top level conversation_

File: data/build.gradle (hunk 6 above)
Comment (agologan) [reply data/build.gradle:32]: Why are there are random spacing changes unrelated to this PR all over the place?
Do a separate change if you want to change the styling.

File: common/src/main/kotlin/com/alfresco/content/PermissionFragment.kt [hunk 11]
```diff
@@ -48,6 +48,25 @@ open class PermissionFragment : Fragment() {
         }
     }
 
+    private fun requestOptionalPermissions(
+        permissions: List<String>,
+        continuation: CancellableContinuation<Boolean>
+    ) {
+        val denied = deniedPermissions(requireContext(), permissions)
+
+        println("PermissionFragment.requestOptionalPermissions $denied")
```
Comment (agologan) [inline common/src/main/kotlin/com/alfresco/content/PermissionFragment.kt:57]: Use `com.alfresco.Logger` if you want to keep the log, instead of `println`

Comment (aman-alfresco) [reply common/src/main/kotlin/com/alfresco/content/PermissionFragment.kt:57]: I already removed this.

Comment (aman-alfresco) [reply common/src/main/kotlin/com/alfresco/content/PermissionFragment.kt:57]: Done

File: browse/src/main/kotlin/com/alfresco/content/browse/BrowseViewState.kt [hunk 12]
```diff
@@ -12,6 +12,7 @@ import com.alfresco.list.merge
 import com.alfresco.list.replace
 import java.time.ZonedDateTime
 import java.time.temporal.ChronoField
+import kotlinx.coroutines.flow.merge
```
Comment (agologan) [inline browse/src/main/kotlin/com/alfresco/content/browse/BrowseViewState.kt:15]: This is an unnecessary import the `merge` reference is `com.alfresco.list.merge`

File: actions/src/main/kotlin/com/alfresco/content/actions/ActionCaptureMedia.kt [hunk 13]
```diff
@@ -6,6 +6,7 @@ import com.alfresco.capture.CaptureHelperFragment
 import com.alfresco.content.PermissionFragment
 import com.alfresco.content.data.Entry
 import com.alfresco.content.data.OfflineRepository
+import java.lang.Exception
```
Comment (agologan) [inline actions/src/main/kotlin/com/alfresco/content/actions/ActionCaptureMedia.kt:9]: This is an unnecessary import the `Exception` reference used here is `com.alfresco.content.actions.Action.Exception`

**Review by agologan — APPROVED** [review]
